hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java

```diff
@@ -298,28 +299,61 @@ public Object internalLoad(
 			// caller did not request forceful eager loading, see if we can create
 			// some form of proxy
 
+			final boolean allowBytecodeProxy = getFactory()
+					.getSessionFactoryOptions()
+					.isEnhancementAsProxyEnabled();
+
+			final boolean entityHasHibernateProxyFactory = persister.getEntityMetamodel()
+					.getTuplizer()
+					.getProxyFactory() != null;
+
 			// first, check to see if we can use "bytecode proxies"
 
 			final EntityMetamodel entityMetamodel = persister.getEntityMetamodel();
 			final BytecodeEnhancementMetadata bytecodeEnhancementMetadata = entityMetamodel.getBytecodeEnhancementMetadata();
 			if ( allowBytecodeProxy && bytecodeEnhancementMetadata.isEnhancedForLazyLoading() ) {
 
-				// we cannot use bytecode proxy for entities with subclasses
-				if ( !entityMetamodel.hasSubclasses() ) {
+				// if the entity defines a HibernateProxy factory, see if there is an
+				// existing proxy associated with the PC - and if so, use it
+				if ( entityHasHibernateProxyFactory ) {
+					final PersistenceContext persistenceContext = getPersistenceContext();
+					final Object proxy = persistenceContext.getProxy( entityKey );
+
+					if ( proxy != null ) {
+						LOG.trace( "Entity proxy found in session cache" );
+
+						final LazyInitializer li = ( (HibernateProxy) proxy ).getHibernateLazyInitializer();
+						if ( li.isUnwrap() ) {
+							return li.getImplementation();
+						}
+
+						return persistenceContext.narrowProxy( proxy, persister, entityKey, null );
+					}
+
+					// specialized handling for entities with subclasses with a HibernateProxy factory
+					if ( persister.getEntityMetamodel().hasSubclasses() ) {
```

Comment:
can be `entityMetamodel.hasSubclasses()`

Reply:
OK, I'll change this as well.